src/core_test/mpas_test_core_halo_exch.F

```diff
@@ -1370,5 +1373,231 @@ subroutine test_core_halo_exch_validate_fields(domain, threadErrs, err)!{{{
+                          intArgs = (/iCell, cellsOnCell(j,iCell)/))
+#endif
+                   exit
+               end if
+            end do
+         end if
+      end do
+#ifdef HALO_EXCH_DEBUG_VERBOSE
+      call mpas_log_write(' mark_interior nCellsMarked:$i sentinal:$i', &
+          intArgs=(/nCellsMarked, sentinelValue/))
+#endif
+
+   end function mark_interior_cells
+
+   !***********************************************************************
+   !> \brief   Identify interior and edge cells
+   !> \Author Michael Duda, Jim Wittig
+   !> \details 
+   !>  This function identifies cells which are on the interior, vs cells on the edge
+   !>  (cells with neighbors which aren't in this block).
+   !>  Returns a vector of flags corresponding to the array of cells, where a value of zero indicates
+   !>  the cell is interior, and a value of non-zero indicates
+   !>    1. the cell is a halo cell (not owned by this block)
+   !>    2. the cell is on an edge of this block (adjacent to a halo cell), or
+   !>    3. the cell is adjacent to an edge cell, or
+   !>    4. the cell is adjacent to a cell which is adjacent to an edge
+   !>  This identifies cells which should be updated via the halo adjoint exchange.
+   !>  Cells marked with a 2 should get updated via halo layer 1.
+   !>  Cells marked with a 3 should get updated via halo layer 2, etc.
+   !>  
+   !-----------------------------------------------------------------------
+   function findExteriorCells(nCellsSolve, nCells, cellsOnCell, edgesOnCell, nHaloLayers) &
+      result(exteriorCells)
+
+      integer, intent(in) :: nCellsSolve !< the number of cells in this block
+      integer, intent(in) :: nCells !< total number of cells (cells in this block plus halo cells)
+      integer, dimension(:,:), intent(in) :: cellsOnCell !< array with adjacent cells for each cell
+      integer, dimension(:), intent(in) :: edgesOnCell !< array with edges for each cell
+      integer, intent(in) :: nHaloLayers  !< the number of halo layers
+
+      integer, dimension(:), allocatable :: exteriorCells
+      integer nInterior, nEdge, nLayers
+
+      allocate(exteriorCells(nCells))
+      exteriorCells(1:nCellsSolve) = 0 !< mark all owned cells as interior
+      exteriorCells(nCellsSolve+1:nCells) = 1 !< mark all halo cells as edge
+      nInterior = 0
+      nEdge = 0
+#ifdef HALO_EXCH_DEBUG_VERBOSE
+      call mpas_log_write(' halo cellsOnCell($i x $i)', &
+         intArgs=(/size(cellsOnCell, dim=1), size(cellsOnCell, dim=2)/))
+#endif
+      
+      ! At this point, only halo cells are marked 1, and all owned cells are marked 0
+      ! for each halo layer, mark cells adjacent to already marked cells with next highest marker
+      do nLayers = 1, nHaloLayers
+         nEdge = nEdge + mark_interior_cells(exteriorCells(1:nCells), nLayers, cellsOnCell, edgesOnCell)
+      end do
+
+      ! Now, interior cells that are adjacent to halo cells are marked 2 (edge cells)
```

Comment:
Do we still need to keep lines 1462 - 1468 ?

Reply:
No, I will remove then.